clang/include/clang/Basic/Attr.td

```diff
@@ -1538,11 +1533,44 @@ def CUDAShared : InheritableAttr {
 }
 def : MutualExclusions<[CUDAConstant, CUDAShared, HIPManaged]>;
 
-def SYCLKernel : InheritableAttr {
-  let Spellings = [Clang<"sycl_kernel">];
-  let Subjects = SubjectList<[FunctionTmpl]>;
-  let LangOpts = [SYCLDevice];
+def DeviceKernel : DeclOrTypeAttr {
+  let Spellings = [Clang<"device_kernel">, Clang<"sycl_kernel">,
+                   Clang<"nvptx_kernel">, Clang<"amdgpu_kernel">,
+                   CustomKeyword<"__kernel">, CustomKeyword<"kernel">];
+  let LangOpts = [];
   let Documentation = [SYCLKernelDocs];
+  let AdditionalMembers =
+  [{
+   inline bool isAMDGPUSpelling() const {
+     return isAMDGPUSpelling(*this);
+   }
+   template<typename T>
+   static inline bool isAMDGPUSpelling(const T& Attr) {
+     return Attr.getAttrName()->getName() == "amdgpu_kernel";
+   }
+   inline bool isNVPTXSpelling() const {
+     return isNVPTXSpelling(*this);
+   }
```

Comment:
There's an easier way to do this, I think. We have:
```
  let Accessors = [Accessor<"isAMDGPU", [Clang<"amdgpu_kernel">]>,
                   Accessor<"isNVPTX", [Clang<"nvptx_kernel">]>];
```
which generates an accessor method for you which returns true/false based on the spelling.

Reply:
Ah cool, thanks! Will do! I think I still need some of the methods to deal with the fact `ParsedAttr` isn't a subclass of `Attr` but also has a `getAttrName()` function, but we can get rid of them

Reply:
I tried this, but we actually need to check the spelling mostly for `ParsedAttr`, so the accessors here won't help. I implemented Erich's idea to use the enum instead of the string. Let me know if you have some tablegen magic that we can use!